### PR TITLE
[Dynamic Buffer][Mellanox] Enhance port shared headroom pool check

### PR DESCRIPTION
**Pull request by stephenxs**
**What I did**

Check maximum port headroom pool in dynamic buffer calculation
Depends on https://github.com/sonic-net/sonic-buildimage/pull/20085

**Why I did it**

To avoid orchagent from crash

**How I verified it**

Manual test
Regression test https://github.com/sonic-net/sonic-mgmt/pull/14479

**Details if related**

### CODE REVIEW

File: cfgmgr/buffer_check_headroom_mellanox.lua [hunk 1]
```diff
@@ -55,17 +59,31 @@ end
...
+
 local lossy_pg_size = pipeline_latency * 1024
 accumulative_size = accumulative_size + lossy_pg_size + egress_mirror_size
 
 -- Fetch all keys in BUFFER_PG according to the port
 redis.call('SELECT', appl_db)
 
+local is_shp_enabled
+local shp_size = tonumber(redis.call('HGET', 'BUFFER_POOL_TABLE:ingress_lossless_pool', 'xoff'))
+if shp_size == nil or shp_size == 0 then
```
Comment (kperumalbfn) [inline cfgmgr/buffer_check_headroom_mellanox.lua:80]: Can we check shared headroom pool based on the mode dynamic or static??

File: cfgmgr/buffer_check_headroom_mellanox.lua [hunk 2]
```diff
@@ -122,26 +140,50 @@ end
...
     end
     if current_profile_size == 0 then
         current_profile_size = lossy_pg_size
     end
     accumulative_size = accumulative_size + current_profile_size * get_number_of_pgs(pg_key)
-    table.insert(debuginfo, 'debug:' .. pg_key .. ':' .. profile .. ':' .. current_profile_size .. ':' .. get_number_of_pgs(pg_key) .. ':accu:' .. accumulative_size)
+
+    if is_shp_enabled and current_profile_xon and current_profile_xoff then
+        if current_profile_size < current_profile_xon + current_profile_xoff then
+            accumulative_shared_headroom = accumulative_shared_headroom + (current_profile_xon + current_profile_xoff - current_profile_size) * get_number_of_pgs(pg_key)
```
Comment (kperumalbfn) [inline cfgmgr/buffer_check_headroom_mellanox.lua:168]: Is this calculated only for active ports?

**Comment by stephenxs** [comment]
Re "Can we check shared headroom pool based on the mode dynamic or static??"
No need to check that. All the Lua plugins are invoked only in dynamic mode

Re "Is this calculated only for active ports?"
Yes

获取 Outlook for iOS<https://aka.ms/o0ukef>
________________________________
发件人: Kumaresh Perumal ***@***.***>
发送时间: 星期三, 九月 25, 2024 8:11 上午
收件人: sonic-net/sonic-swss ***@***.***>
抄送: Stephen Sun ***@***.***>; Author ***@***.***>
主题: Re: [sonic-net/sonic-swss] [Dynamic Buffer][Mellanox] Enhance port shared headroom pool check (PR #3272)


@kperumalbfn commented on this pull request.

________________________________

In cfgmgr/buffer_check_headroom_mellanox.lua<https://github.com/sonic-net/sonic-swss/pull/3272#discussion_r1774240901>:

>  local lossy_pg_size = pipeline_latency * 1024
 accumulative_size = accumulative_size + lossy_pg_size + egress_mirror_size

 -- Fetch all keys in BUFFER_PG according to the port
 redis.call('SELECT', appl_db)

+local is_shp_enabled
+local shp_size = tonumber(redis.call('HGET', 'BUFFER_POOL_TABLE:ingress_lossless_pool', 'xoff'))
+if shp_size == nil or shp_size == 0 then


Can we check shared headroom pool based on the mode dynamic or static??

________________________________

In cfgmgr/buffer_check_headroom_mellanox.lua<https://github.com/sonic-net/sonic-swss/pull/3272#discussion_r1774241686>:

>      end
     if current_profile_size == 0 then
         current_profile_size = lossy_pg_size
     end
     accumulative_size = accumulative_size + current_profile_size * get_number_of_pgs(pg_key)
-    table.insert(debuginfo, 'debug:' .. pg_key .. ':' .. profile .. ':' .. current_profile_size .. ':' .. get_number_of_pgs(pg_key) .. ':accu:' .. accumulative_size)
+
+    if is_shp_enabled and current_profile_xon and current_profile_xoff then
+        if current_profile_size < current_profile_xon + current_profile_xoff then
+            accumulative_shared_headroom = accumulative_shared_headroom + (current_profile_xon + current_profile_xoff - current_profile_size) * get_number_of_pgs(pg_key)


Is this calculated only for active ports?

―

File: cfgmgr/buffer_check_headroom_mellanox.lua (hunk 1 above)
Comment (stephenxs) [reply cfgmgr/buffer_check_headroom_mellanox.lua:80]: No need to check that. All the Lua plugins are invoked only in dynamic mode

File: cfgmgr/buffer_check_headroom_mellanox.lua (hunk 2 above)
Comment (stephenxs) [reply cfgmgr/buffer_check_headroom_mellanox.lua:168]: Yes

**Review by kperumalbfn — APPROVED** [review]
LGTM

**Comment by stephenxs** [comment]
The failures do not look relevant to the PR. Retrying.

**Review by prsunny — APPROVED** [review]

**Pull request merged**